// Job/Job.js
var pc = require('ProcessChain');

function JavaScript(func)
{
    if (typeof func === "function")
        this.exec = func;
    this._next = undefined;
    this._data = undefined;
    this._written = false;
}

// Should be overridden
JavaScript.prototype.exec = function(data)
{
    return data;
};

JavaScript.prototype.chain = function(js)
{
    var obj = this;
    while (obj._next !== undefined)
        obj = obj._next;
    obj._next = js;
};

JavaScript.prototype.write = function(data)
{
    var r = this.exec(data);
    if (this._next)
        this._next.write(r);
    else
        this._data = r;
    this._written = true;
};

JavaScript.prototype.end = function(callback)
{
    var obj = this;

    if (!this._written)
        this.write();

    while (obj._next !== undefined)
        obj = obj._next;
    callback(obj._data);
    obj._data = undefined;
};

function Job()
{
    this._jobs = [];
    this._chains = [];
}

Job.prototype.process = function(process)
{
    var idx = this._jobs.length;
    if (this._jobs.length === 0 || this._jobs[idx - 1].type !== "process")
        this._jobs.push({type: "process", jobs: [process]});
    else
        this._jobs[idx - 1].jobs.push(process);
    return this;
};

Job.prototype.proc = Job.prototype.process;

Job.prototype.js = function(js)
{
    if (!(js instanceof JavaScript))
        return this;
    js._next = undefined;

    var idx = this._jobs.length;
    if (this._jobs.length === 0 || this._jobs[idx - 1].type !== "js")
        this._jobs.push({type: "js", job: js});
    else
        this._jobs[idx - 1].job.chain(js);
    return this;
};

Job.prototype.exec = function(callback)
{
    var pchain = undefined;
    var jchain = undefined;
    var i, j;
    for (i in this._jobs) {
        var job = this._jobs[i];
        if (job.type === "process") {
            pchain = new pc.ProcessChain(global.jsh.jshNative);
            for (j in job.jobs) {
                pchain.chain(job.jobs[j]);
            }
            if (jchain) {
                jchain.end(function(data) { pchain.write(data); });
                jchain = undefined;
            }
        } else {
            jchain = job.job;
            if (pchain) {
                pchain.end(function(data) { jchain.write(data); });
                pchain = undefined;
            }
        }
    }
    if (pchain) {
        pchain.end(callback);
    } else if (jchain) {
        jchain.end(callback);
    }
};

module.exports = {
    Job: Job,
    JavaScript: JavaScript
};
